# ¡Hola, Sebastián!  

Mi nombre es Carlos Ortiz, soy code reviewer de TripleTen y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión. 

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>
¡Empecemos!

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df, uso el método pd.read_csv con la ruta indicada.
df = pd.read_csv("/datasets/music_project_en.csv")

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con la importación de datos y de librerías.
</div>

Muestra las 10 primeras filas de la tabla:

In [3]:
#Uso el método Head() con el parámetro 10.
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Uso el metodo .info() para examinar los datos generales sobre la tabla
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


In [5]:
#Revisar los valores NA detectados en el paso anterior
df.isna().sum()

  userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64

In [6]:
#Revisar Duplicados con el método Duplicated().sum() 
df.duplicated().sum()

3826

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `'userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `UserID` debe escribirse en snake_case.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

**"Hipótesis: La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad."**

El Data Frame contiene información suficiente para dar respuesta a la hipótesis, sin embargo se deben primero procesar los datos, ya que gracias al método .info() notamos que hay valores ausentes, y con el método duplicated() que hay valores duplicados. Es importante hacer una limpieza antes de poder analizar los datos, además de verificar si hay posibles duplicados implícitos.

Los tipos de datos son strings para todas las columnas, estos representan información descriptiva acerca de las canciones, los usuarios y los dias cuando se reprodujeron de la siguiente manera:

- Las columnas `'Track'`, `'artist'`, `'genre'`: información de la canción reproducida.
- Las columnas: `'userID'`  & `'city'`: información del usuario y la cuidad donde se encontraba.
- Las columnas `'time'` & `'Day'`: hora y dia de reproducción de las canciones.

Valores ausentes por columna: `'artist'` es la columna con mas Null.
- `'artist'`      7567
- `'Track'`       1343
- `'genre'`       1198


Filas duplicadas para todo el data set:
- 3826


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy buen trabajo con esta exploración de datos y con la lectura inicial de los problemas.
</div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [7]:
# uso el atributo .columns para ver el nombre de las columnas
df.columns


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [8]:
# Bucle en los encabezados poniendo todo en minúsculas

columns_lower = []                     # creo una lista vacía para los nombre nuevos

for col in df.columns:                 # creo un for que recorra las columnas del data frame
    name_lower = col.lower()           # utilizo el método  .lower() y lo asigno a una nueva var.
    columns_lower.append(name_lower)   # utilizo el metodo append() para los valores nuevos
    
df.columns = columns_lower             # asigno la lista de nombres a las columnas del data frame
df.columns                             # verifico que todos los nombres esten en minúsculas: OK

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [9]:
# Bucle en los encabezados eliminando los espacios

columns_strip = []                     # creo una lista vacía para los nombre nuevos

for col in df.columns:                 # creo un for que recorra las columnas del data frame
    x = col.strip()                    # utilizo el método .strip() y lo asigno a una nueva var.
    columns_strip.append(x)            # utilizo el método append() para los valores nuevos
    
df.columns = columns_strip             # asigno la lista de nombres a las columnas del data frame
df.columns                             # verifico los nombres estén libres de espacios: OK


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [10]:
# creo un diccionario con el cambio deseado, utilizo el método rename() con el parámetro inplace

new_user_id = {"userid" : "user_id"}   
df.rename(columns = new_user_id, inplace = True)
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Comprueba el resultado. Muestra los encabezados una vez más:

In [11]:
# las columnas ahora están listas: en minúsculas, sin espacios y user_id en snake_case
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Has renombrado las columnas correctamente.
</div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [12]:
# uso métodos en cadena: isna() y .sum()
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [13]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

columns_to_repalce = ['track', 'artist', 'genre']  # creo una lista con las columnas con Null
for col in columns_to_repalce:                     # creo un bucle que recorra las columnas 
    df[col].fillna("unknown", inplace = True)      # utilizo el metodo .fillna()

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
¡Perfecto!
</div>

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [14]:
# Confirmar que ya no hay valores ausentes: OK
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [15]:
# uso métodos en cadena: .duplicated() y .sum()
df.duplicated().sum()

3826

In [16]:
num = "{:.0%}".format(3826/65079)
print (f"% de valores duplicados: {num}") # 3826 de 65079 datos. use format para reducir los decimales

% de valores duplicados: 6%


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [17]:
# Eliminar duplicados explícitos usando el método drop.duplicates()
df = df.drop_duplicates().reset_index(drop = True)
df.head(3)

user_id                        track            artist genre         city  \
0  FFB692EC            Kamigata To Boots  The Mass Missile  rock  Shelbyville   
1  55204538  Delayed Because of Accident  Andreas Rönnberg  rock  Springfield   
2    20EC38            Funiculì funiculà       Mario Lanza   pop  Shelbyville   

       time        day  
0  20:28:33  Wednesday  
1  14:07:09     Friday  
2  20:58:07  Wednesday

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [18]:
# Confirmar que no hay valores duplicados:
df.duplicated().sum()

0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [19]:
# Inspeccionar los nombres de géneros únicos

genre_unq = df["genre"].unique()  # uso em método unique() en la columna genre
genre_unq.sort()                  # ordeno el resultado con el método sort()
genre_unq                         # muestro los valores para el análisis

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [20]:
# Función para reemplazar duplicados implícitos

def replace_wrong_genres(wrong_genres,correct_genre):          #creo la funcion con dos parámetros
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre,correct_genre)   #uso .replace()
    return df                                                           

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [21]:
# Eliminar duplicados implícitos

wrong_genres = ["hip","hop","hip-hop"]  # Creo una lista con los duplicados implícitos encontrados
correct_genre = "hiphop"                # Valor correcto para remplazar    

x =  replace_wrong_genres(wrong_genres,correct_genre)   # llamo la función

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [22]:
# Comprobación de duplicados implícitos

genre_unq = df["genre"].unique()         # uso el método unique() en la columna genre
genre_unq.sort()                           # ordeno el resultado con el método sort()
genre_unq                                # muestro los valores para el análisis

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

Durante el análisis de duplicados se encontró que existían 3826 duplicados explícitos que equivalían a el 6% de los datos originales, estos se eliminaron para mejorar le precisión del análisis.


Posteriormente se analizó la columna de `'genres'`, indispensable para nuestro análisis en la cual se identifico la presencia de duplicados impliciticos para el género “Hip-Hop”. como esto también puede ocurrir con otros géneros, se creó una función que permite remplazar los valores incorrectos de forma automatizada para resolver posibles casos similares.


El análisis de duplicados junto con el de valores ausentes permite asegurar que se hará un análisis con más precisión, que permita aceptar o refutar nuestra hipótesis.


[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [23]:
# Contar las canciones reproducidas en cada ciudad

df.groupby(by= "city")["track"].count() # agrupar por Ciudad y usar count() en la columna "track"



city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

In [24]:
num = "{0:.2g}".format(42741/18512) #cuantas reproduciones más en Spr que en She
print (f"{num} veces más en Springfield")

2.3 veces más en Springfield


**Observación:**

En Springfield se escuchan 2.3 veces más canciones que Shelbyville.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [25]:
# Calcular las canciones reproducidas en cada uno de los tres días

day_sort = df.groupby(by= "day")["track"].count()  # contar los valores en "track" por día
day_sort.sort_values()                             # ordenar los valores
day_sort

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

**Observación:** 

Viernes es el día que más se escucha música, miércoles es el día con menos reproducciones.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [26]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(day,city):
    
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtered_day = df[df["day"] == day]
    
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filtered_city = filtered_day[filtered_day["city"] == city]
    
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    total = filtered_city['user_id'].count()
    
    # Devolve el número de valores de la columna 'user_id'
    return total

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [27]:
df.head(3) #Uso head para poder ver los valores con mayor facilidad

user_id                        track            artist genre         city  \
0  FFB692EC            Kamigata To Boots  The Mass Missile  rock  Shelbyville   
1  55204538  Delayed Because of Accident  Andreas Rönnberg  rock  Springfield   
2    20EC38            Funiculì funiculà       Mario Lanza   pop  Shelbyville   

       time        day  
0  20:28:33  Wednesday  
1  14:07:09     Friday  
2  20:58:07  Wednesday

In [28]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks("Monday","Springfield")

15740

In [29]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks("Monday","Shelbyville")

5614

In [30]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks("Wednesday","Springfield")

11056

In [31]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks("Wednesday","Shelbyville")

7003

In [32]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks("Friday","Springfield")

15945

In [33]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks("Friday","Shelbyville")

5895

In [34]:
# otra forma de ver los resultados de nuestra funcion es con el método .pivot_table:
# ya que permite definir la cuidad como filas, el dia como columnas y podemos pedir que cuente el número de canciones
# este devuelve un data frame nuevo y permite confirmar que nuestra funcion devuelve los resultados deseados

pivot = df[["day", "city", "track"]]
pd.pivot_table(pivot,index = "city", columns = "day" ,aggfunc = "count")


track                 
day         Friday Monday Wednesday
city                               
Shelbyville   5895   5614      7003
Springfield  15945  15740     11056

**Conclusiones**

Hipótesis: La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


**Con los datos podemos afirmar para cada una de las cuidades:**



> 1) en **Springfield** se escuchan **más canciones los lunes y los viernes**: ambos dias con más de 15000 reproducciones.


> 2) Por el contrario, en **shelbyville** el dia con más reproducciones es **el miércoles**: con 7000, y presenta menos reproducciones los lunes y viernes.

lo cual indicaría que la hipótesis es correcta


[Volver a Contenidos](#back)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
¡Excelente trabajo en esta sección!
</div>

# Conclusiones <a id='end'></a>

Podemos entonces concluir que la actividad de los usuarios si difiere según el dia de la semana y la cuidad, el análisis muestra incrementos en las reproducciones para dias específicos además de que este comportamiento es diferente en ambas ciudades, respaldando la hipótesis.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-success">
    
# Comentarios generales
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Has aprobado un nuevo proyecto. ¡Felicitaciones!
</div>

[Volver a Contenidos](#back)